File: code/seungwookim99/week8/prog_42579.py
def solution(genres, plays):
    answer = []
    n = len(genres)
    genres_distinct = list(set(genres))
    total_plays = {g: 0 for g in genres_distinct}
    play_list = [] #(장르 총 플레이 수, 해당 노래 플레이 수, 고유 번호, genre)
    for i in range(n):
        total_plays[genres[i]] += plays[i]
    for i in range(n):
        genre = genres[i]
        play_list.append((total_plays[genre], plays[i], i, genre))
    play_list.sort(key=lambda x: (-x[0], -x[1], x[2]))
    
    curr_genre = ""
    cnt = 0
    for p in play_list:
        if curr_genre == p[3]:
            if cnt == 2:
                continue
            else:
                answer.append(p[2])
                cnt += 1
        else:
            curr_genre = p[3]
            answer.append(p[2])
            cnt = 1
    return answer

File: code/kauthenticity/week8/prog_42579.js
function solution(genres, plays) {
    var answer = [];
    const playCntPerGenre = {}; // [장르]: 플레이 횟수
    const playCntPerSong = {}; // [장르]: [횟수, 번호]
    const n = genres.length;

    genres.forEach((genre, i) => {
        if (playCntPerSong[genre]) {
            playCntPerSong[genre].push([plays[i], i]);
        } else {
            playCntPerSong[genre] = [[plays[i], i]];
        }

        if (playCntPerGenre[genre]) {
            playCntPerGenre[genre] += plays[i];
        } else {
            playCntPerGenre[genre] = plays[i];
        }
    });

    Object.keys(playCntPerSong).forEach((key) => {
        playCntPerSong[key].sort(([cnt1, idx1], [cnt2, idx2]) => {
            if (cnt1 == cnt2) {
                return idx1 - idx2;
            }
            return cnt2 - cnt1;
        });
    });

    Object.entries(playCntPerGenre)
        .sort(([_, cnt1], [__, cnt2]) => {
            return cnt2 - cnt1;
        })
        .forEach(([genre]) => {
            if (playCntPerSong[genre].length == 1) {
                answer.push(playCntPerSong[genre][0][1]);
            } else {
                answer.push(
                    playCntPerSong[genre][0][1],
                    playCntPerSong[genre][1][1]
                );
            }
        });

    return answer;
}
